README
NAME
    `Net::Async::CassandraCQL' - use Cassandra databases with IO::Async
    using CQL

SYNOPSIS
     use IO::Async::Loop;
     use Net::Async::CassandraCQL;
     use Protocol::CassandraCQL qw( CONSISTENCY_QUORUM );

     my $loop = IO::Async::Loop->new;

     my $cass = Net::Async::CassandraCQL->new(
        host => "localhost",
        keyspace => "my-keyspace",
        default_consistency => CONSISTENCY_QUORUM,
     );
     $loop->add( $cass );

     $cass->connect->get;

     my @f;
     foreach my $number ( 1 .. 100 ) {
        push @f, $cass->query( "INSERT INTO numbers (v) VALUES $number" );
     }
     Future->needs_all( @f )->get;

     my $get_stmt = $cass->prepare( "SELECT v FROM numbers" )->get;

     my ( undef, $result ) = $get_stmt->execute( [] )->get;

     foreach my $row ( $result->rows_hash ) {
        say "We have a number " . $row->{v};
     }

DESCRIPTION
    This module allows use of the `CQL3' interface of a Cassandra database.
    It fully supports asynchronous operation via IO::Async, allowing both
    direct queries and prepared statements to be managed concurrently, if
    required. Alternatively, as the interface is entirely based on Future
    objects, it can be operated synchronously in a blocking fashion by
    simply awaiting each individual operation by calling the `get' method.

    It is based on Protocol::CassandraCQL, which more completely documents
    the behaviours and limits of its ability to communicate with Cassandra.

EVENTS
  on_node_up $nodeid
  on_node_down $nodeid
    The node's status has changed. `$nodeid' is the node's IP address as a
    text string.

  on_node_new $nodeid
  on_node_removed $nodeid
    A new node has been added to the cluster, or an existing node has been
    decommissioned and removed.

    These four events are obtained from event watches on the actual node
    connections and filtered to remove duplicates. The use of multiple
    primaries should improve the reliability of notifications, though if
    multiple nodes fail at or around the same time this may go unreported,
    as no node will ever report its own failure.

PARAMETERS
    The following named parameters may be passed to `new' or `configure':

    host => STRING
    hosts => ARRAY of STRING
            The hostnames of Cassandra node to connect to initially. If more
            than one host is provided in an array, they will be attempted
            sequentially until one succeeds during the intial connect phase.

    service => STRING
            Optional. The service name or port number to connect to.

    username => STRING
    password => STRING
            Optional. Authentication details to use for
            `PasswordAuthenticator'.

    keyspace => STRING
            Optional. If set, a `USE keyspace' query will be issued as part
            of the connect method.

    default_consistency => INT
            Optional. Default consistency level to use if none is provided
            to `query' or `execute'.

    primaries => INT
            Optional. The number of primary node connections to maintain.
            Defaults to 1 if not specified.

    prefer_dc => STRING
            Optional. If set, prefer to pick primary nodes from the given
            data center, only falling back on others if there are not enough
            available.

    cql_version => INT
            Optional. Version of the CQL wire protocol to negotiate during
            connection. Defaults to 1.

METHODS
  $str = $cass->quote( $str )
    Quotes a string argument suitable for inclusion in an immediate CQL
    query string.

    In general, it is better to use a prepared query and pass the value as
    an execute parameter though.

  $str = $cass->quote_identifier( $str )
    Quotes an identifier name suitable for inclusion in a CQL query string.

  $cass->connect( %args ) ==> ()
    Connects to the Cassandra node and starts up the connection. The
    returned Future will yield nothing on success.

    Takes the following named arguments:

    host => STRING
    hosts => ARRAY of STRING
    service => STRING

    A set of host names are required, either as a named argument or as a
    configured value on the object. If the service name is missing, the
    default CQL port will be used instead.

  $cass->close_when_idle ==> $cass
    Stops accepting new queries and prepares all the existing connections to
    be closed once every outstanding query has been responded to. Returns a
    future that will eventually yield the CassandraCQL object, when all the
    connections are closed.

    After calling this method it will be an error to invoke `query',
    `prepare', `execute' or the various other methods derived from them.

  $cass->close_now
    Immediately closes all node connections and shuts down the object. Any
    outstanding or queued queries will immediately fail. Consider this as a
    "last resort" failure shutdown, as compared to the graceful draining
    behaviour of `close_when_idle'.

  $cass->query( $cql, $consistency, %other_args ) ==> ( $type, $result )
    Performs a CQL query. On success, the values returned from the Future
    will depend on the type of query.

    For `USE' queries, the type is `keyspace' and `$result' is a string
    giving the name of the new keyspace.

    For `CREATE', `ALTER' and `DROP' queries, the type is `schema_change'
    and `$result' is a 3-element ARRAY reference containing the type of
    change, the keyspace and the table name.

    For `SELECT' queries, the type is `rows' and `$result' is an instance of
    Protocol::CassandraCQL::Result containing the returned row data.

    For other queries, such as `INSERT', `UPDATE' and `DELETE', the future
    returns nothing.

    `%other_args' may be any of the following, when using `cql_version' 2 or
    above:

    skip_metadata => BOOL
            Requests the server does not include result metadata in the
            response. It will be up to the caller to provide this, via
            `set_metadata' on the returned Result object, before it can be
            used.

    page_size => INT
            Requests that the server returns at most the given number of
            rows. If any further remain, the result object will include the
            `paging_state' field. This can be passed in another `query' call
            to obtain the next set of data.

    paging_state => INT
            Requests that the server continues a paged request from this
            position, given in a previous response.

    serial_consistency => INT
            Sets the consistency level for serial operations in the query.
            Must be one of `CONSISTENCY_SERIAL' or
            `CONSISTENCY_LOCAL_SERIAL'.

  $cass->query_rows( $cql, $consistency, %other_args ) ==> $result
    A shortcut wrapper for `query' which expects a `rows' result and returns
    it directly. Any other result is treated as an error. The returned
    Future returns a `Protocol::CassandraCQL::Result' directly

  $cass->prepare( $cql ) ==> $query
    Prepares a CQL query for later execution. On success, the returned
    Future yields an instance of a prepared query object (see below).

    Query objects stored internally cached by the CQL string; subsequent
    calls to `prepare' with the same exact CQL string will yield the same
    object immediately, saving a roundtrip.

  $cass->execute( $query, $data, $consistency, %other_args ) ==> ( $type, $result )
    Executes a previously-prepared statement, given the binding data. On
    success, the returned Future will yield results of the same form as the
    `query' method. `$data' should contain a list of encoded byte-string
    values.

    Normally this method is not directly required - instead, use the
    `execute' method on the query object itself, as this will encode the
    parameters correctly.

    `%other_args' may be as for the `query' method.

CONVENIENT WRAPPERS
    The following wrapper methods all wrap the basic `query' operation.

  $cass->schema_keyspaces ==> $result
    A shortcut to a `SELECT' query on `system.schema_keyspaces', which
    returns a result object listing all the keyspaces.

    Exact details of the returned columns will depend on the Cassandra
    version, but the result should at least be keyed by the first column,
    called `keyspace_name'.

     my $keyspaces = $result->rowmap_hash( "keyspace_name" )

  $cass->schema_columnfamilies( $keyspace ) ==> $result
    A shortcut to a `SELECT' query on `system.schema_columnfamilies', which
    returns a result object listing all the columnfamilies of the given
    keyspace.

    Exact details of the returned columns will depend on the Cassandra
    version, but the result should at least be keyed by the first column,
    called `columnfamily_name'.

     my $columnfamilies = $result->rowmap_hash( "columnfamily_name" )

  $cass->schema_columns( $keyspace, $columnfamily ) ==> $result
    A shortcut to a `SELECT' query on `system.schema_columns', which returns
    a result object listing all the columns of the given columnfamily.

    Exact details of the returned columns will depend on the Cassandra
    version, but the result should at least be keyed by the first column,
    called `column_name'.

     my $columns = $result->rowmap_hash( "column_name" )

TODO
    *       Allow other load-balancing strategies than roundrobin.

    *       Adjust connected primary nodes when changing `primaries'
            parameter.

    *       Allow backup nodes, for faster connection failover.

    *       Support LZ4 compression when using CQL version 2.

            This is blocked on RT #92825

SPONSORS
    This code was paid for by

    * Perceptyx http://www.perceptyx.com/

    * Shadowcat Systems http://www.shadow.cat

AUTHOR
    Paul Evans <leonerd@leonerd.org.uk>



Changes
Revision history for Net-Async-CassandraCQL

0.11    2014/06/01 12:35:56
        [CHANGES]
         * Optionally use LZ4 compression when using CQLv2

        [BUGFIXES]
         * Avoid relying on strong forward references in Future, by creating
           intentional cycles on pending Futures. Workaround for bugfix in
           upcoming Future release.

0.10    2014/02/07 22:28:24
        [CHANGES]
         * Allow a list of initial seed hosts for connect reliability
         * Use TOPOLOGY_CHANGE events to keep the nodelist updated
         * Support CQLv2
         * Make Snappy compression optional

0.09    2013/09/25 18:31:37
        [CHANGES]
         * Allow unreferenced prepared queries to live for a few more
           minutes before expiring them, in case of re-prepare

0.08    2013/09/24 23:08:45
        [CHANGES]
         * When a preferred node comes UP, switch to it if we have a
           connection to non-preferred node
         * Cache query objects by CQL string; return identical objects
           immediately rather than re-preparing duplicate queries

0.07    2013/09/18 18:20:45
        [CHANGES]
         * Allow connecting to multiple nodes, optionally with a preference
           on which data center to pick nodes from
         * Watch for STATUS_CHANGE events to keep node list updated on which
           nodes are up or down
         * Derive directly from IO::Async::Stream, not IaProtocol::Stream
         * Added some IaNotifier ->debug_printf calls

0.06    2013/09/12 17:39:26
        [CHANGES]
         * First attempt at clustering support - NaCassandraCQL object now
           represents the cluster as a whole, with individual connections
         * Query the system.peers table at connect time, and fail-over to
           another node if the connection fails

0.05    2013/09/03 17:08:58
        [CHANGES]
         * Support password-based authentication
         * Support registering interest in server events, and invoking a
           callback then they arrive

0.04    CHANGES:
         * Moved Protocol::CassandraCQL to its own distribution
         * Added ->query_rows, ->quote, ->quote_identifier
         * Added some convenience methods to query schema metadata
           (schema_keyspaces, schema_columnfamilies, schema_columns)

0.03    CHANGES:
         * Allow setting a default consistency level to avoid needing one at
           every call to ->query or ->execute
         * Validate perl data being passed to ->execute bind parameters
         * Added $result->rowmap_array and ->rowmap_hash
         * Implement INET type
         * Moved more parsing/building code into Protocol::CassandraCQL to
           prepare for distribution split

0.02    CHANGES:
         * Added $result->rows_array and ->rows_hash
         * Represent types as firstclass objects
         * Implement collection types (LIST, SET, MAP)
         * Implement UUID and TIMEUUID types
         * Updated documentation

        BUGFIXES:
         * Account for inexactness of DOUBLE result printing in unit tests
         * Abort build if no 64bit pack format support
         * Make Protocol::CassandraCQL::Frame perl 5.8.8-clean

0.01    First version, released on an unsuspecting world.



lib/Net/Async/CassandraCQL/Connection.pm
#  You may distribute under the terms of either the GNU General Public License
#  or the Artistic License (the same terms as Perl itself)
#
#  (C) Paul Evans, 2013-2014 -- leonerd@leonerd.org.uk

package Net::Async::CassandraCQL::Connection;

use strict;
use warnings;
use 5.010;

our $VERSION = '0.11';

use base qw( IO::Async::Stream );
IO::Async::Stream->VERSION( '0.59' );

use Carp;

use Future 0.13;

use constant HAVE_SNAPPY => eval { require Compress::Snappy };
use constant HAVE_LZ4    => eval { require Compress::LZ4 };

use Protocol::CassandraCQL qw(
   :opcodes :results :consistencies FLAG_COMPRESS
   build_frame parse_frame
);
use Protocol::CassandraCQL::Frame;
use Protocol::CassandraCQL::Frames 0.10 qw( :all );

use Net::Async::CassandraCQL::Query;

# Ensure that IO::Async definitely uses this for Iv6 connections as we need
# the ->peerhost method
require IO::Socket::IP;

=head1 NAME

C<Net::Async::CassandraCQL::Connection> - connect to a single Cassandra database node

=head1 DESCRIPTION

TODO

=cut

=head1 EVENTS

=head2 on_event $name, @args

A registered event occurred. C<@args> will depend on the event name. Each
is also available as its own event, with the name in lowercase. If the event
is not one of the types recognised below, C<@args> will contain the actual
L<Protocol::CassandraCQL::Frame> object.

=head2 on_topology_change $type, $node

The cluster topology has changed. C<$node> is a packed socket address.

=head2 on_status_change $status, $node

The node's status has changed. C<$node> is a packed socket address.

=head2 on_schema_change $type, $keyspace, $table

A keyspace or table schema has changed.

=cut

=head1 PARAMETERS

The following named parameters may be passed to C<new> or C<configure>:

=over 8

=item username => STRING

=item password => STRING

Optional. Authentication details to use for C<PasswordAuthenticator>.

=item cql_version => INT

Optional. Version of the CQL wire protocol to negotiate during connection.
Defaults to 1.

=back

=cut

sub _init
{
   my $self = shift;
   $self->SUPER::_init( @_ );

   $self->{streams} = []; # map [1 .. 127] to Future
   $self->{pending} = []; # queue of [$opcode, $frame, $f]
   $self->{cql_version} = 1;
}

sub configure
{
   my $self = shift;
   my %params = @_;

   foreach (qw( username password cql_version
                on_event on_topology_change on_status_change on_schema_change )) {
      $self->{$_} = delete $params{$_} if exists $params{$_};
   }

   $self->SUPER::configure( %params );
}

=head1 METHODS

=cut

=head2 $id = $conn->nodeid

Returns the connection's node ID (the string form of its IP address), which is
used as its ID in the C<system.peers> table.

=cut

sub nodeid
{
   my $self = shift;
   $self->{nodeid};
}

sub _version
{
   my $self = shift;
   return $self->{cql_version};
}

# function
sub _decode_result
{
   my ( $version, $response ) = @_;

   my ( $type, $result ) = parse_result_frame( $version, $response );

   if( $type == RESULT_VOID ) {
      return Future->new->done();
   }
   else {
      if   ( $type == RESULT_ROWS          ) { $type = "rows"          }
      elsif( $type == RESULT_SET_KEYSPACE  ) { $type = "keyspace"      }
      elsif( $type == RESULT_SCHEMA_CHANGE ) { $type = "schema_change" }
      return Future->new->done( $type => $result );
   }
}

=head2 $conn->connect( %args ) ==> $conn

Connects to the Cassandra node an send the C<OPCODE_STARTUP> message. The
returned Future will yield the connection itself on success.

Takes the following named arguments:

=over 8

=item host => STRING

=item service => STRING

=back

=cut

sub connect
{
   my $self = shift;
   my %args = @_;

   $args{socktype} ||= "stream";

   return ( $self->{connect_f} ||=
      $self->SUPER::connect( %args )->on_fail( sub { undef $self->{connect_f} } ) )
      ->then( sub {
         $self->{nodeid} = $self->read_handle->peerhost;
         $self->startup
      })->then( sub { Future->new->done( $self ) });
}

sub _has_pending
{
   my $self = shift;
   defined and return 1 for @{ $self->{streams} };
   return 0;
}

sub on_read
{
   my $self = shift;
   my ( $buffref, $eof ) = @_;

   my ( $version, $flags, $streamid, $opcode, $body ) = parse_frame( $$buffref ) or return 0;

   $version & 0x80 or
      $self->fail_all_and_close( "Expected response to have RESPONSE bit set" ), return;
   $version &= 0x7f;

   # Test version <= for now in case of "unsupported protocol version" error messages, and
   # test it exactly later
   $version <= $self->_version or
      $self->fail_all_and_close( sprintf "Unexpected message version %#02x\n", $version ), return;

   if( $flags & FLAG_COMPRESS and my $decompress = $self->{decompress} ) {
      $flags &= ~FLAG_COMPRESS;
      $body = $decompress->( $body );
   }

   $flags == 0 or
      $self->fail_all_and_close( sprintf "Unexpected message flags %#02x\n", $flags ), return;

   my $frame = Protocol::CassandraCQL::Frame->new( $body );

   if( my $f = $self->{streams}[$streamid] ) {
      undef $self->{streams}[$streamid];

      if( $opcode == OPCODE_ERROR ) {
         my ( $err, $message ) = parse_error_frame( $version, $frame );
         $f->fail( "OPCODE_ERROR: $message\n", $err, $frame );
      }
      else {
         $version == $self->_version or
            $self->fail_all_and_close( sprintf "Unexpected message version %#02x\n", $version ), return;

         $f->done( $opcode, $frame, $version );
      }

      if( my $next = shift @{ $self->{pending} } ) {
         my ( $opcode, $frame, $f ) = @$next;
         $self->_send( $opcode, $streamid, $frame, $f );
      }
      elsif( my $close_f = $self->{cassandra_close_future} and !$self->_has_pending ) {
         $close_f->done( $self );
      }
   }
   elsif( $streamid == 0 and $opcode == OPCODE_ERROR ) {
      my ( $err, $message ) = parse_error_frame( $version, $frame );
      $self->fail_all_and_close( "OPCODE_ERROR: $message\n", $err, $frame );
   }
   elsif( $streamid == 0xff and $opcode == OPCODE_EVENT ) {
      $self->_event( $frame );
   }
   else {
      print STDERR "Received a message opcode=$opcode for unknown stream $streamid\n";
   }

   return 1;
}

sub _event
{
   my $self = shift;
   my ( $frame ) = @_;

   my ( $name, @args ) = parse_event_frame( $self->_version, $frame );

   $self->maybe_invoke_event( "on_".lc($name), @args )
      or $self->maybe_invoke_event( on_event => $name, @args );
}

sub on_closed
{
   my $self = shift;
   $self->fail_all( "Connection closed" );
}

sub fail_all
{
   my $self = shift;
   my ( $failure ) = @_;

   foreach ( @{ $self->{streams} } ) {
      $_->fail( $failure ) if $_;
   }
   @{ $self->{streams} } = ();

   foreach ( @{ $self->{pending} } ) {
      $_->[2]->fail( $failure );
   }
   @{ $self->{pending} } = ();
}

sub fail_all_and_close
{
   my $self = shift;
   my ( $failure ) = @_;

   $self->fail_all( $failure );

   $self->close;

   return Future->new->fail( $failure );
}

=head2 $conn->send_message( $opcode, $frame ) ==> ( $reply_opcode, $reply_frame, $reply_version )

Sends a message with the given opcode and L<Protocol::CassandraCQL::Frame> for
the message body. The returned Future will yield the response opcode, frame
and version number (with the RESPONSE bit masked off).

This is a low-level method; applications should instead use one of the wrapper
methods below.

=cut

sub send_message
{
   my $self = shift;
   my ( $opcode, $frame ) = @_;

   croak "Cannot ->send_message when in close-pending state" if $self->{cassandra_close_future};

   my $f = $self->loop->new_future;

   my $streams = $self->{streams} ||= [];
   my $id;
   foreach ( 1 .. $#$streams ) {
      $id = $_ and last if !defined $streams->[$_];
   }

   if( !defined $id ) {
      if( $#$streams == 127 ) {
         push @{ $self->{pending} }, [ $opcode, $frame, $f ];
         return $f;
      }
      $id = @$streams;
      $id = 1 if !$id; # can't use 0
   }

   $self->_send( $opcode, $id, $frame, $f );

   return $f;
}

sub _send
{
   my $self = shift;
   my ( $opcode, $id, $frame, $f ) = @_;

   my $flags = 0;
   my $body = $frame->bytes;

   if( my $compress = $self->{compress} ) {
      my $body_compressed = $compress->( $body );
      if( length $body_compressed < length $body ) {
         $flags |= FLAG_COMPRESS;
         $body = $body_compressed;
      }
   }

   $self->write( build_frame( $self->_version, $flags, $id, $opcode, $body ) );

   $self->{streams}[$id] = $f;
}

=head2 $conn->startup ==> ()

Sends the initial connection setup message. On success, the returned Future
yields nothing.

Normally this is not required as the C<connect> method performs it implicitly.

=cut

sub startup
{
   my $self = shift;

   # CQLv1 doesn't support LZ4
   my $compression;
   if( HAVE_LZ4 and $self->_version > 1 ) {
      # Cassandra prepends 32bit BE integer of original size
      $compression = [ lz4 =>
         sub { my ( $data ) = @_; pack "N a*", length $data, Compress::LZ4::lz4_compress( $data ) },
         sub { my ( $bodylen, $lz4data ) = unpack "N a*", $_[0]; Compress::LZ4::lz4_decompress( $lz4data, $bodylen ) },
      ];
   }
   elsif( HAVE_SNAPPY ) {
      $compression = [ snappy =>
         \&Compress::Snappy::compress,
         \&Compress::Snappy::decompress,
      ];
   }
   # else undef

   my $f = $self->send_message( OPCODE_STARTUP, build_startup_frame( $self->_version,
      options => {
         CQL_VERSION => "3.0.5",
         ( $compression ? ( COMPRESSION => $compression->[0] ) : () ),
      } )
   )->then( sub {
      my ( $op, $response, $version ) = @_;

      if( $op == OPCODE_READY ) {
         return Future->new->done;
      }
      elsif( $op == OPCODE_AUTHENTICATE ) {
         return $self->_authenticate( parse_authenticate_frame( $version, $response ) );
      }
      else {
         return $self->fail_all_and_close( "Expected OPCODE_READY or OPCODE_AUTHENTICATE" );
      }
   });

   $self->{compress}   = $compression->[1];
   $self->{decompress} = $compression->[2];

   return $f;
}

sub _authenticate
{
   my $self = shift;
   my ( $authenticator ) = @_;

   if( $authenticator eq "org.apache.cassandra.auth.PasswordAuthenticator" ) {
      foreach (qw( username password )) {
         defined $self->{$_} or croak "Cannot authenticate by password without $_";
      }

      $self->send_message( OPCODE_CREDENTIALS, build_credentials_frame( $self->_version,
         credentials => {
            username => $self->{username},
            password => $self->{password},
         } )
      )->then( sub {
         my ( $op, $response, $version ) = @_;
         $op == OPCODE_READY or return $self->fail_all_and_close( "Expected OPCODE_READY" );

         return Future->new->done;
      });
   }
   else {
      return $self->fail_all_and_close( "Unrecognised authenticator $authenticator" );
   }
}

=head2 $conn->options ==> $options

Requests the list of supported options from the server node. On success, the
returned Future yields a HASH reference mapping option names to ARRAY
references containing valid values.

=cut

sub options
{
   my $self = shift;

   $self->send_message( OPCODE_OPTIONS,
      Protocol::CassandraCQL::Frame->new
   )->then( sub {
      my ( $op, $response, $version ) = @_;
      $op == OPCODE_SUPPORTED or return Future->new->fail( "Expected OPCODE_SUPPORTED" );

      my ( $opts ) = parse_supported_frame( $version, $response );
      return Future->new->done( $opts );
   });
}

=head2 $conn->query( $cql, $consistency, %other_args ) ==> ( $type, $result )

Performs a CQL query. On success, the values returned from the Future will
depend on the type of query.

For C<USE> queries, the type is C<keyspace> and C<$result> is a string giving
the name of the new keyspace.

For C<CREATE>, C<ALTER> and C<DROP> queries, the type is C<schema_change> and
C<$result> is a 3-element ARRAY reference containing the type of change, the
keyspace and the table name.

For C<SELECT> queries, the type is C<rows> and C<$result> is an instance of
L<Protocol::CassandraCQL::Result> containing the returned row data.

For other queries, such as C<INSERT>, C<UPDATE> and C<DELETE>, the future
returns nothing.

Any other arguments will be passed on to the underlying C<build_query_frame>
function of L<Protocol::CassandraCQL::Frames>.

=cut

sub query
{
   my $self = shift;
   my ( $cql, $consistency, %other_args ) = @_;

   $self->send_message( OPCODE_QUERY, build_query_frame( $self->_version,
         cql         => $cql,
         consistency => $consistency,
         %other_args,
      )
   )->then( sub {
      my ( $op, $response, $version ) = @_;
      $op == OPCODE_RESULT or return Future->new->fail( "Expected OPCODE_RESULT" );
      return _decode_result( $version, $response );
   });
}

=head2 $conn->prepare( $cql ) ==> $query

Prepares a CQL query for later execution. On success, the returned Future
yields an instance of L<Net::Async::CassandraCQL::Query>.

=cut

sub prepare
{
   my $self = shift;
   my ( $cql, $cassandra ) = @_;

   $self->send_message( OPCODE_PREPARE, build_prepare_frame( $self->_version,
         cql => $cql,
      )
   )->then( sub {
      my ( $op, $response, $version ) = @_;
      $op == OPCODE_RESULT or return Future->new->fail( "Expected OPCODE_RESULT" );

      my ( $type, $result ) = parse_result_frame( $version, $response );
      $type == RESULT_PREPARED or return Future->new->fail( "Expected RESULT_PREPARED" );

      my ( $id, $params_meta, $result_meta ) = @$result;

      my $query = Net::Async::CassandraCQL::Query->new(
         cassandra   => $cassandra,
         cql         => $cql,
         id          => $id,
         params_meta => $params_meta,
         result_meta => $result_meta, # v2+ only
      );
      return Future->new->done( $query );
   });
}

=head2 $conn->execute( $id, $data, $consistency, %other_args ) ==> ( $type, $result )

Executes a previously-prepared statement, given its ID and the binding data.
On success, the returned Future will yield results of the same form as the
C<query> method. C<$data> should contain a list of encoded byte-string values.
Any other arguments will be passed on to the underlying C<build_execute_frame>
function of L<Protocol::CassandraCQL::Frames>.

Normally this method is not directly required - instead, use the C<execute>
method on the query object itself, as this will encode the parameters
correctly.

=cut

sub execute
{
   my $self = shift;
   my ( $id, $data, $consistency, %other_args ) = @_;

   $self->send_message( OPCODE_EXECUTE, build_execute_frame( $self->_version,
         id          => $id,
         values      => $data,
         consistency => $consistency,
         %other_args,
      )
   )->then( sub {
      my ( $op, $response, $version ) = @_;
      $op == OPCODE_RESULT or return Future->new->fail( "Expected OPCODE_RESULT" );
      return _decode_result( $version, $response );
   });
}

=head2 $conn->register( $events ) ==> ()

Registers the connection's interest in receiving events of the types given in
the ARRAY reference. Event names may be C<TOPOLOGY_CHANGE>, C<STATUS_CHANGE>
or C<SCHEMA_CHANGE>. On success, the returned Future yields nothing.

=cut

sub register
{
   my $self = shift;
   my ( $events ) = @_;

   $self->send_message( OPCODE_REGISTER, build_register_frame( $self->_version,
         events => $events,
      )
   )->then( sub {
      my ( $op, $response, $version ) = @_;
      $op == OPCODE_READY or Future->new->fail( "Expected OPCODE_READY" );

      return Future->new->done;
   });
}

=head2 $conn->close_when_idle ==> $conn

If the connection is idle (has no outstanding queries), then it is closed
immediately. If not, it is put into close-pending mode, where it will accept
no more queries, and will close when the last pending one is complete.

Returns a future which will eventually yield the (closed) connection when it
becomes closed.

=cut

sub close_when_idle
{
   my $self = shift;

   $self->{cassandra_close_future} ||= do {
      my $f = $self->loop->new_future;
      $f->on_done( sub { $_[0]->close } );

      $f->done( $self ) if !$self->_has_pending;

      $f
   };
}

=head1 SPONSORS

This code was paid for by

=over 2

=item *

Perceptyx L<http://www.perceptyx.com/>

=item *

Shadowcat Systems L<http://www.shadow.cat>

=back

=head1 AUTHOR

Paul Evans <leonerd@leonerd.org.uk>

=cut

0x55AA;
